<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Vignette package rdadapt v1.0 • rdadapt</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Vignette package rdadapt v1.0">
<meta property="og:description" content="rdadapt">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rdadapt</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/vignette_tutorial.html">
    <span class="fas fa-stethoscope"></span>
     
    tutorial
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-book fa-lg"></span>
     
    Reference
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../articles/news.html">
    <span class="fas fa-newspaper"></span>
     
    News
  </a>
</li>
<li>
  <a href="https://github.com/landscape-genomics/rdadapt" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Vignette package rdadapt v1.0</h1>
                        <h4 data-toc-skip class="author">Thibaut
Capblancq &amp; Maya Gueguen</h4>
            
            <h4 data-toc-skip class="date">2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/landscape-genomics/rdadapt/blob/HEAD/vignettes/vignette_tutorial.Rmd" class="external-link"><code>vignettes/vignette_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>vignette_tutorial.Rmd</code></div>

    </div>

    
    
<p><br></p>
<p>In the following, we show how the <code>rdadapt</code> package can
perform:</p>
<ul>
<li>genome scans for selection,</li>
<li>adaptive index projection on the landscape,</li>
<li>estimation of genomic offsets</li>
<li>identification of co-varying modules of adaptive loci or</li>
<li>identification of adaptive groups of individual or populations.</li>
</ul>
<p>We show how to run the package using data simulated with Slim3 as
described in <a href="https://doi.org/10.1093/molbev/msad140" class="external-link">Gain et
al. 2023</a>.</p>
<p>The <code>maladapt</code> dataset contains genotypes (<em>Y</em>) for
2036 simulated individuals on a 12*12 grid constrained by two causal
environmental variables (<em>Var1</em> and <em>Var2</em>). Fitness
(<em>fitness</em>) was measured for all individuals before and after a
brutal change in environment.</p>
<p>To run the package, you need to install it (and load it) using the
following command lines:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"landscape-genomics/rdadapt"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://landscape-genomics.github.io/rdadapt/" class="external-link">rdadapt</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Other libraries essentially used for plotting purposes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/valentint/robust" class="external-link">robust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.sthda.com/english/rpkgs/factoextra" class="external-link">factoextra</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Do we have to keep those here?</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/jdstorey/qvalue" class="external-link">qvalue</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vegandevs/vegan" class="external-link">vegan</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RevolutionAnalytics/foreach" class="external-link">foreach</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/plyr" class="external-link">plyr</a></span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<h2>
<ol style="list-style-type: decimal"><li>
<p>Data Preparation</p>
</li></ol>
</h2>

<p><br></p>
<p><code>maladapt</code> dataset contains informations about individuals
and populations, and corresponding genetic informations.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"maladapt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">maladapt</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "pop"        "coord"      "Y"          "xx.current" "xx.future" </span></span>
<span><span class="co">## [6] "causal1"    "causal2"    "fitness"</span></span></code></pre>
<p><br></p>
<h4>
1.1 Metadata
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Population labels</span></span>
<span><span class="va">pop</span> <span class="op">&lt;-</span> <span class="va">maladapt</span><span class="op">$</span><span class="va">pop</span></span>
<span></span>
<span><span class="co">## Individual coordinates</span></span>
<span><span class="va">coord</span> <span class="op">&lt;-</span> <span class="va">maladapt</span><span class="op">$</span><span class="va">coord</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Genotypes</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">maladapt</span><span class="op">$</span><span class="va">Y</span></span>
<span><span class="va">Y</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Current and future climate values</span></span>
<span><span class="va">xx.current</span> <span class="op">&lt;-</span> <span class="va">maladapt</span><span class="op">$</span><span class="va">xx.current</span></span>
<span><span class="va">xx.future</span> <span class="op">&lt;-</span> <span class="va">maladapt</span><span class="op">$</span><span class="va">xx.future</span></span>
<span></span>
<span><span class="co">## Rasterize the variables</span></span>
<span><span class="va">xx.ras.current</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span>, time <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span>
<span>                                  , y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span>, each <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span>
<span>                                  , <span class="fu"><a href="https://rspatial.github.io/terra/reference/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">xx.current</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="va">coord</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span>
<span>                                                                    , <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="va">coord</span><span class="op">$</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>                                              , <span class="va">mean</span><span class="op">)</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>                       , type <span class="op">=</span> <span class="st">"xyz"</span></span>
<span>                       , crs <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/crs.html" class="external-link">crs</a></span><span class="op">(</span><span class="st">"epsg:4326"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xx.ras.future</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span>, time <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span>
<span>                                 , y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span>, each <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span>
<span>                                 , <span class="fu"><a href="https://rspatial.github.io/terra/reference/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">xx.future</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="va">coord</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span>
<span>                                                                  , <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="va">coord</span><span class="op">$</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>                                           , <span class="va">mean</span><span class="op">)</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>                      , type <span class="op">=</span> <span class="st">"xyz"</span></span>
<span>                      , crs <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/crs.html" class="external-link">crs</a></span><span class="op">(</span><span class="st">"epsg:4326"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
<h4>
1.2 Sub-sample populations
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Randomly select only 25 populations to mimic an empirical dataset</span></span>
<span><span class="va">samp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">pop</span> <span class="op"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">25</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Y.samp</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">[</span><span class="va">samp</span>, <span class="op">]</span></span>
<span><span class="va">xx.current.samp</span> <span class="op">&lt;-</span> <span class="va">xx.current</span><span class="op">[</span><span class="va">samp</span>, <span class="op">]</span></span>
<span><span class="va">xx.future.samp</span> <span class="op">&lt;-</span> <span class="va">xx.future</span><span class="op">[</span><span class="va">samp</span>, <span class="op">]</span></span>
<span><span class="va">coord.samp</span> <span class="op">&lt;-</span> <span class="va">coord</span><span class="op">[</span><span class="va">samp</span>, <span class="op">]</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-10-1.png" width="576" style="display: block; margin: auto;"></p>
<p><br></p>
<h2>
<ol start="2" style="list-style-type: decimal"><li>
<p>Genome Scan</p>
</li></ol>
</h2>

<p><br></p>
<h4>
2.1 Identifying loci under selection using RDA
</h4>
<p>The <code>rdadapt</code> function performs redundancy analysis and
computes p-values to test for outliers based on loci extremeness along a
distribution of Mahalanobis distances estimated between each locus and
the center of the RDA space using a certain number of axes (K).</p>
<p>The function accommodates individual genotypes or allele
frequencies.</p>
<p>The first step of the procedure is to regress the genetic matrix
against the environmental predictors using a RDA model.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RDA_env</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cca.html" class="external-link">rda</a></span><span class="op">(</span><span class="va">Y</span> <span class="op">~</span> <span class="va">.</span>,  <span class="va">xx.current</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">RDA_env</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We then have to choose a number of RDA axes to include when
conducting the genome scan.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/screeplot.html" class="external-link">screeplot</a></span><span class="op">(</span><span class="va">RDA_env</span>, main <span class="op">=</span> <span class="st">"Eigenvalues of constrained axes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-12-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Looking at the proportion of variance explained by each axis, we
decided to use the first two axes.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Running the function with K = 2</span></span>
<span><span class="va">rdadapt_env</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rdadapt.html">rdadapt</a></span><span class="op">(</span>RDA <span class="op">=</span> <span class="va">RDA_env</span>, K <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>One critical step when conducting a genome scan is to set a pertinent
p-value threshold to identify the outlier loci. Here, we used a
<em>Bonferroni correction</em> to account for multiple testing.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## P-values threshold after Bonferroni correction</span></span>
<span><span class="va">thres_env</span> <span class="op">&lt;-</span> <span class="fl">0.01</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">rdadapt_env</span><span class="op">$</span><span class="va">p.values</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Identifying the loci that are below the p-value threshold</span></span>
<span><span class="va">outliers</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">rdadapt_env</span><span class="op">$</span><span class="va">p.values</span> <span class="op">&lt;</span> <span class="va">thres_env</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">outliers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 215</span></span></code></pre>
<p>The function found 215 outliers.</p>
<p>Once the outliers have been identified, it can be useful to visualize
their distribution in comparison with neutral loci using either a RDA
biplot.</p>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-15-1.png" width="576" style="display: block; margin: auto;"></p>
<p><br></p>
<h2>
<ol start="3" style="list-style-type: decimal"><li>
<p>Adaptive Index</p>
</li></ol>
</h2>

<p><br></p>
<h4>
3.1 Adaptively enriched genetic space
</h4>
<p>The 215 putative adaptive loci can be used as multivariate response
in a new “adaptively enriched” RDA, using the same environmental
predictors as explanatory variables.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Adaptively enriched RDA</span></span>
<span><span class="va">RDA_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cca.html" class="external-link">rda</a></span><span class="op">(</span><span class="va">Y</span><span class="op">[</span>, <span class="va">outliers</span><span class="op">]</span> <span class="op">~</span> <span class="va">.</span>,  <span class="va">xx.current</span><span class="op">)</span></span></code></pre></div>
<p>A RDA biplot allows us to visualize the relationship between the
outlier loci and the underlying environmental predictors.</p>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-17-1.png" width="480" style="display: block; margin: auto;"><strong>Note</strong> the RDA space is slightly different here than
above because only the outlier loci were used to train the model. The
goal being to remove any potential noise coming from non-adaptive loci
but if we are unsure of the outlier identification we can use the model
including all loci instead.</p>
<p><br></p>
<h4>
3.2 Adaptive index across the landscape
</h4>
<p>The <code>adaptive_index</code> function will use the scores of the
environmental variables along the RDA axes to calculate a genetic-based
index of adaptation for each environmental pixel of the landscape. This
index is estimated independently for each RDA axis of interest using the
formula: <span class="math display">\[
\sum_{i = 1}^{n}{a_ib_i}
\]</span> Where <em>a</em> is the climatic variable score (loading)
along the RDA axis, <em>b</em> is the standardized value for this
particular variable at the focal pixel, and <em>i</em> refers to one of
the <em>n</em> different variables used in the RDA model.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Running the function for all pixels </span></span>
<span><span class="va">res_RDA_proj_current</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adaptive_index.html">adaptive_index</a></span><span class="op">(</span>RDA <span class="op">=</span> <span class="va">RDA_outliers</span></span>
<span>                                       , K <span class="op">=</span> <span class="fl">2</span></span>
<span>                                       , env <span class="op">=</span> <span class="va">xx.ras.current</span></span>
<span>                                       , env_mask <span class="op">=</span> <span class="cn">NULL</span></span>
<span>                                       , method <span class="op">=</span> <span class="st">"loadings"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>adaptive_index</code> function thus provides an estimate of
adaptive genetic similarity or difference of all pixels on the landscape
as a function of the values of the environmental predictors at that
location. When projected on a map it allows visualizing the different
adaptive gradients across a species range.</p>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-19-1.png" width="960" style="display: block; margin: auto;"></p>
<p><br></p>
<h2>
<ol start="4" style="list-style-type: decimal"><li>
<p>Genomic Offset</p>
</li></ol>
</h2>

<p><br></p>
<h4>
4.1 Predicting population future maladaptation
</h4>
<p>Once the genetic ~ environment relationship is characterized, it can
be extrapolated to future environments to predict a potential shift in
adaptive optimum induced by climate change using the
<code>genomic_offset</code> function.The RDA-based method to predict
this future maladaptation is relatively simple. As done above, RDA can
be used to predict the optimal adaptive genetic composition for each
environmental site under consideration, using both current and future
environmental conditions. The difference between the two predictions
provides an estimate of the change in genetic composition that would be
required to track climate change.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Projecting the adaptive landscape after the change in environment </span></span>
<span><span class="va">res_RDA_proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/genomic_offset.html">genomic_offset</a></span><span class="op">(</span>RDA <span class="op">=</span> <span class="va">RDA_outliers</span></span>
<span>                               , K <span class="op">=</span> <span class="fl">2</span></span>
<span>                               , env_pres <span class="op">=</span> <span class="va">xx.ras.current</span></span>
<span>                               , env_fut <span class="op">=</span> <span class="va">xx.ras.future</span></span>
<span>                               , env_mask <span class="op">=</span> <span class="cn">NULL</span></span>
<span>                               , method <span class="op">=</span> <span class="st">"loadings"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Verifying that prediction are matching with a decline in fitness on the landscape</span></span>
<span><span class="va">fitness</span> <span class="op">&lt;-</span> <span class="va">maladapt</span><span class="op">$</span><span class="va">fitness</span></span>
<span><span class="va">fitness.future</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span>, time <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span>
<span>                                  , y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span>, each <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span>
<span>                                  , <span class="fu"><a href="https://rspatial.github.io/terra/reference/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">fitness</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="va">coord</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span>
<span>                                                                     , <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="va">coord</span><span class="op">$</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>                                              , <span class="va">mean</span><span class="op">)</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>                       , type <span class="op">=</span> <span class="st">"xyz"</span></span>
<span>                       , crs <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/crs.html" class="external-link">crs</a></span><span class="op">(</span><span class="st">"epsg:4326"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-23-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The genomic offset predictions are effectively negatively correlated
with the decrease in fitness that followed a brutal change in
environment.</p>
<p><br></p>
<p>The <code>genomic_offset</code> function can also produce predictions
for specific populations instead of projecting on a landscape. To do so
the user just has to feed the function with a <code>data.frame</code>
instead of a <code>RasterStack</code> as below.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Estimates genomic offset for specific populations</span></span>
<span><span class="va">offset_samp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/genomic_offset.html">genomic_offset</a></span><span class="op">(</span>RDA <span class="op">=</span> <span class="va">RDA_outliers</span></span>
<span>                              , K <span class="op">=</span> <span class="fl">2</span></span>
<span>                              , env_pres <span class="op">=</span> <span class="va">xx.current.samp</span></span>
<span>                              , env_fut <span class="op">=</span> <span class="va">xx.future.samp</span></span>
<span>                              , method <span class="op">=</span> <span class="st">"loadings"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-25-1.png" width="960" style="display: block; margin: auto;"></p>
<p><br></p>
<h4>
4.2. Predicting geographic genomic offset
</h4>
<p>A challenge associated with any predictive method is to find ways to
validate the predictions. Here, we estimated a spatial genomic offset
between the climate of each source population and the climate of a
hypothetical garden: cell[8,1].</p>
<p>The <code>genomic_offset</code> function can estimate spatial offset
when the <code>env_gar</code> argument is used with a simple
<code>vector</code> of environmental values.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Estimates genomic offset for a transplantation experiment</span></span>
<span><span class="va">offset_gar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/genomic_offset.html">genomic_offset</a></span><span class="op">(</span>RDA <span class="op">=</span> <span class="va">RDA_outliers</span></span>
<span>                             , K <span class="op">=</span> <span class="fl">2</span></span>
<span>                             , env_pres <span class="op">=</span> <span class="va">xx.current</span></span>
<span>                             , env_gar <span class="op">=</span> <span class="va">xx.current</span><span class="op">[</span><span class="fl">100</span>, <span class="op">]</span></span>
<span>                             , method <span class="op">=</span> <span class="st">"loadings"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-27-1.png" width="960" style="display: block; margin: auto;"></p>
<p>Based on the adaptively enriched RDA space, we estimated a genomic
offset between each source individual and the common garden climatic
conditions (black cross).</p>
<p>This geographic genomic offset can then be compared to different
fitness traits in the garden to estimate the impact of this predicted
maladaptation on seedling fitness. We can also compare the influence of
genomic offset on fitness with the impact of a basic climate transfer
distance estimated between each source population and the garden from
the climatic variables only (using Mahalanobis distance).</p>
<p><br></p>
<h2>
<ol start="5" style="list-style-type: decimal"><li>
<p>Loci Modules</p>
</li></ol>
</h2>

<p><br></p>
<p>RDA models can also be used to find covarying sets of loci using the
<code>loci_modules</code> function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Groups the outlier loci into two modules</span></span>
<span><span class="va">modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loci_modules.html">loci_modules</a></span><span class="op">(</span>RDA <span class="op">=</span> <span class="va">RDA_outliers</span>, nb_clusters <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-29-1.png" width="480" style="display: block; margin: auto;"></p>
<p>This can be useful to identify group of loci or genomic regions that
respond to different adaptive constraints or are involved in different
adaptive gradients.</p>
<p><br></p>
<h2>
<ol start="6" style="list-style-type: decimal"><li>
<p>Adaptive Groups</p>
</li></ol>
</h2>

<p><br></p>
<p>The <code>adaptive_groups</code> function allows the user to find
groups of covarying individuals or populations based on their scores in
the RDA space and a hkmean clustering procedure.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## New RDA model with only sample populations and outlier loci</span></span>
<span><span class="va">RDA_samp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cca.html" class="external-link">rda</a></span><span class="op">(</span><span class="va">Y.samp</span><span class="op">[</span>,<span class="va">outliers</span><span class="op">]</span> <span class="op">~</span> <span class="va">.</span>,  <span class="va">xx.current.samp</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Finds N optimal groups among the individuals used to build the adaptive RDA space</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adaptive_groups.html">adaptive_groups</a></span><span class="op">(</span>RDA <span class="op">=</span> <span class="va">RDA_samp</span>, K <span class="op">=</span> <span class="fl">2</span>, nb_clusters <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-31-1.png" width="480" style="display: block; margin: auto;"></p>
<p><code>adaptive_groups</code> accommodates discrete populations as
above, or a <code>RasterStack</code> as below.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Finds N groups on the adaptive landscape projected from the RDA model</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adaptive_groups.html">adaptive_groups</a></span><span class="op">(</span>RDA <span class="op">=</span> <span class="va">RDA_samp</span>, K <span class="op">=</span> <span class="fl">2</span>, env <span class="op">=</span> <span class="va">xx.ras.current</span>, nb_clusters <span class="op">=</span> <span class="fl">9</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_tutorial_files/figure-html/unnamed-chunk-33-1.png" width="480" style="display: block; margin: auto;"></p>
<p>The adaptive landscape is continuous so it looks more like
discretizing a continuous distribution. The adaptive indices or genomic
offset described above are probably more appropriate here. However, this
functionality can become handy for management or conservation purposes,
for example to identify conservation units or seed zones across the
range of a species.</p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://scholar.google.fr/citations?user=BuqUlRcAAAAJ&amp;hl=fr&amp;oi=ao" class="external-link">Thibaut Capblancq</a>, <a href="https://mayagueguen.github.io" class="external-link">Maya Gueguen</a>, <a href="https://brenna-forester.weebly.com" class="external-link">Brenna Forester</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
